 # <font color='orange'>Aprendizaje no Supervisado (Parte II)</font>

## **1.** Realizar un análisis exploratorio de los datos para comprender las características y la distribución de los datos.

In [1]:
# Importar libreria para cargar dataset

import pandas as pd

df= pd.read_excel("Telco-Customer-Churn.xlsx")
df.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges  Churn  
0           Electronic check          29.85         29.85     No  
1               Mailed check          56.95       1889.50     No  
2               Mailed check          53.85        108.15    Yes  
3  Bank transfer (automatic)          42.30       1840.75     No  
4           Electronic check          70.70        151.65    Yes  

[5 rows x 21 columns]

In [3]:
# Revisar estructura de datos

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


In [5]:
# Resumen estadístico de los datos

df.describe()

SeniorCitizen       tenure  MonthlyCharges  TotalCharges
count    7043.000000  7043.000000     7043.000000   7032.000000
mean        0.162147    32.371149       64.761692   2283.300441
std         0.368612    24.559481       30.090047   2266.771362
min         0.000000     0.000000       18.250000     18.800000
25%         0.000000     9.000000       35.500000    401.450000
50%         0.000000    29.000000       70.350000   1397.475000
75%         0.000000    55.000000       89.850000   3794.737500
max         1.000000    72.000000      118.750000   8684.800000

**Conclusión del análisis descriptivo**

A partir del análisis exploratorio de las variables numéricas del conjunto de datos, se pueden extraer las siguientes conclusiones:

**SeniorCitizen:**

El 16% de los clientes son adultos mayores (mean = 0.16).

Es una variable binaria: 0 indica que no es adulto mayor y 1 que sí.

La mayoría de los clientes no son adultos mayores, lo que podría tener impacto en los patrones de cancelación.

**Tenure (tiempo como cliente):**

El valor mínimo es 0 meses y el máximo 72 meses.

La mediana está en 29 meses, lo que indica que al menos la mitad de los clientes ha estado suscrito por más de 2 años y medio.

La distribución es amplia, lo que sugiere que hay tanto clientes nuevos como clientes de larga duración.

**MonthlyCharges (cargos mensuales):**

Varía entre 18.25 y 118.75, con un promedio de 64.76.

La desviación estándar de 30 sugiere variabilidad considerable en los planes contratados.

Algunos clientes pagan muy poco y otros más del doble del promedio.

**TotalCharges (cargos totales):**

El rango es amplio: desde 18.8 hasta más de 8600.

La media es 2283.30, con una alta desviación estándar, indicando que algunos clientes han sido muy rentables para la empresa.

Hay 11 valores faltantes, probablemente asociados a clientes muy nuevos con tenure = 0.

## **2.** Preprocesar los datos, incluyendo la limpieza, la transformación y la normalización de las características según sea necesario.

**Objetivos del preprocesamiento:**

**a-.** Limpieza de datos

**b-.** Transformación de variables categóricas

**c-.** Conversión de tipos de datos

**d-.** Normalización (escalado) de variables numéricas

**e-.** Separar la variable objetivo (Churn)

In [10]:
# a-. Limpieza de datos
# Ver cantidad de valores nulos por columna
df.isnull().sum()


customerID           0
gender               0
SeniorCitizen        0
Partner              0
Dependents           0
tenure               0
PhoneService         0
MultipleLines        0
InternetService      0
OnlineSecurity       0
OnlineBackup         0
DeviceProtection     0
TechSupport          0
StreamingTV          0
StreamingMovies      0
Contract             0
PaperlessBilling     0
PaymentMethod        0
MonthlyCharges       0
TotalCharges        11
Churn                0
dtype: int64

In [12]:
# Eliminar filas con valores nulos en la columna TotalCharges
df = df.dropna(subset=['TotalCharges'])

# Verificar que ya no hay valores nulos
df.isnull().sum()

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

In [14]:
# b-. Transformación de variables categóricas

#Eliminamos customerID es solo un identidicador único, no aporta valor predictivo.
df = df.drop('customerID', axis=1) 
df

gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0     Female              0     Yes         No       1           No   
1       Male              0      No         No      34          Yes   
2       Male              0      No         No       2          Yes   
3       Male              0      No         No      45           No   
4     Female              0      No         No       2          Yes   
...      ...            ...     ...        ...     ...          ...   
7038    Male              0     Yes        Yes      24          Yes   
7039  Female              0     Yes        Yes      72          Yes   
7040  Female              0     Yes        Yes      11           No   
7041    Male              1     Yes         No       4          Yes   
7042    Male              0      No         No      66          Yes   

         MultipleLines InternetService OnlineSecurity OnlineBackup  \
0     No phone service             DSL             No          Yes   
1                   No             DSL            Yes           No   
2                   No             DSL            Yes          Yes   
3     No phone service             DSL            Yes           No   
4                   No     Fiber optic             No           No   
...                ...             ...            ...          ...   
7038               Yes             DSL            Yes           No   
7039               Yes     Fiber optic             No          Yes   
7040  No phone service             DSL            Yes           No   
7041               Yes     Fiber optic             No           No   
7042                No     Fiber optic            Yes           No   

     DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0                  No          No          No              No  Month-to-month   
1                 Yes          No          No              No        One year   
2                  No          No          No              No  Month-to-month   
3                 Yes         Yes          No              No        One year   
4                  No          No          No              No  Month-to-month   
...               ...         ...         ...             ...             ...   
7038              Yes         Yes         Yes             Yes        One year   
7039              Yes          No         Yes             Yes        One year   
7040               No          No          No              No  Month-to-month   
7041               No          No          No              No  Month-to-month   
7042              Yes         Yes         Yes             Yes        Two year   

     PaperlessBilling              PaymentMethod  MonthlyCharges  \
0                 Yes           Electronic check           29.85   
1                  No               Mailed check           56.95   
2                 Yes               Mailed check           53.85   
3                  No  Bank transfer (automatic)           42.30   
4                 Yes           Electronic check           70.70   
...               ...                        ...             ...   
7038              Yes               Mailed check           84.80   
7039              Yes    Credit card (automatic)          103.20   
7040              Yes           Electronic check           29.60   
7041              Yes               Mailed check           74.40   
7042              Yes  Bank transfer (automatic)          105.65   

      TotalCharges Churn  
0            29.85    No  
1          1889.50    No  
2           108.15   Yes  
3          1840.75    No  
4           151.65   Yes  
...            ...   ...  
7038       1990.50    No  
7039       7362.90    No  
7040        346.45    No  
7041        306.60   Yes  
7042       6844.50    No  

[7032 rows x 20 columns]

In [16]:
# Codificar columnas binarias (Yes/No) usando Label Encoding

# c-. Conversión de tipos de datos

columnas_binarias = ['Partner', 'Dependents', 'PhoneService', 'PaperlessBilling', 'Churn']
df[columnas_binarias] = df[columnas_binarias].replace({'Yes': 1, 'No': 0}) 

df

C:\Users\Niscar\AppData\Local\Temp\ipykernel_16300\1240734031.py:6: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df[columnas_binarias] = df[columnas_binarias].replace({'Yes': 1, 'No': 0})


gender  SeniorCitizen  Partner  Dependents  tenure  PhoneService  \
0     Female              0        1           0       1             0   
1       Male              0        0           0      34             1   
2       Male              0        0           0       2             1   
3       Male              0        0           0      45             0   
4     Female              0        0           0       2             1   
...      ...            ...      ...         ...     ...           ...   
7038    Male              0        1           1      24             1   
7039  Female              0        1           1      72             1   
7040  Female              0        1           1      11             0   
7041    Male              1        1           0       4             1   
7042    Male              0        0           0      66             1   

         MultipleLines InternetService OnlineSecurity OnlineBackup  \
0     No phone service             DSL             No          Yes   
1                   No             DSL            Yes           No   
2                   No             DSL            Yes          Yes   
3     No phone service             DSL            Yes           No   
4                   No     Fiber optic             No           No   
...                ...             ...            ...          ...   
7038               Yes             DSL            Yes           No   
7039               Yes     Fiber optic             No          Yes   
7040  No phone service             DSL            Yes           No   
7041               Yes     Fiber optic             No           No   
7042                No     Fiber optic            Yes           No   

     DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0                  No          No          No              No  Month-to-month   
1                 Yes          No          No              No        One year   
2                  No          No          No              No  Month-to-month   
3                 Yes         Yes          No              No        One year   
4                  No          No          No              No  Month-to-month   
...               ...         ...         ...             ...             ...   
7038              Yes         Yes         Yes             Yes        One year   
7039              Yes          No         Yes             Yes        One year   
7040               No          No          No              No  Month-to-month   
7041               No          No          No              No  Month-to-month   
7042              Yes         Yes         Yes             Yes        Two year   

      PaperlessBilling              PaymentMethod  MonthlyCharges  \
0                    1           Electronic check           29.85   
1                    0               Mailed check           56.95   
2                    1               Mailed check           53.85   
3                    0  Bank transfer (automatic)           42.30   
4                    1           Electronic check           70.70   
...                ...                        ...             ...   
7038                 1               Mailed check           84.80   
7039                 1    Credit card (automatic)          103.20   
7040                 1           Electronic check           29.60   
7041                 1               Mailed check           74.40   
7042                 1  Bank transfer (automatic)          105.65   

      TotalCharges  Churn  
0            29.85      0  
1          1889.50      0  
2           108.15      1  
3          1840.75      0  
4           151.65      1  
...            ...    ...  
7038       1990.50      0  
7039       7362.90      0  
7040        346.45      0  
7041        306.60      1  
7042       6844.50      0  

[7032 rows x 20 columns]

In [18]:
# Codificar el género

df['gender'] = df['gender'].replace({'Female': 1, 'Male': 0})
df


C:\Users\Niscar\AppData\Local\Temp\ipykernel_16300\634632398.py:3: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df['gender'] = df['gender'].replace({'Female': 1, 'Male': 0})


gender  SeniorCitizen  Partner  Dependents  tenure  PhoneService  \
0          1              0        1           0       1             0   
1          0              0        0           0      34             1   
2          0              0        0           0       2             1   
3          0              0        0           0      45             0   
4          1              0        0           0       2             1   
...      ...            ...      ...         ...     ...           ...   
7038       0              0        1           1      24             1   
7039       1              0        1           1      72             1   
7040       1              0        1           1      11             0   
7041       0              1        1           0       4             1   
7042       0              0        0           0      66             1   

         MultipleLines InternetService OnlineSecurity OnlineBackup  \
0     No phone service             DSL             No          Yes   
1                   No             DSL            Yes           No   
2                   No             DSL            Yes          Yes   
3     No phone service             DSL            Yes           No   
4                   No     Fiber optic             No           No   
...                ...             ...            ...          ...   
7038               Yes             DSL            Yes           No   
7039               Yes     Fiber optic             No          Yes   
7040  No phone service             DSL            Yes           No   
7041               Yes     Fiber optic             No           No   
7042                No     Fiber optic            Yes           No   

     DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0                  No          No          No              No  Month-to-month   
1                 Yes          No          No              No        One year   
2                  No          No          No              No  Month-to-month   
3                 Yes         Yes          No              No        One year   
4                  No          No          No              No  Month-to-month   
...               ...         ...         ...             ...             ...   
7038              Yes         Yes         Yes             Yes        One year   
7039              Yes          No         Yes             Yes        One year   
7040               No          No          No              No  Month-to-month   
7041               No          No          No              No  Month-to-month   
7042              Yes         Yes         Yes             Yes        Two year   

      PaperlessBilling              PaymentMethod  MonthlyCharges  \
0                    1           Electronic check           29.85   
1                    0               Mailed check           56.95   
2                    1               Mailed check           53.85   
3                    0  Bank transfer (automatic)           42.30   
4                    1           Electronic check           70.70   
...                ...                        ...             ...   
7038                 1               Mailed check           84.80   
7039                 1    Credit card (automatic)          103.20   
7040                 1           Electronic check           29.60   
7041                 1               Mailed check           74.40   
7042                 1  Bank transfer (automatic)          105.65   

      TotalCharges  Churn  
0            29.85      0  
1          1889.50      0  
2           108.15      1  
3          1840.75      0  
4           151.65      1  
...            ...    ...  
7038       1990.50      0  
7039       7362.90      0  
7040        346.45      0  
7041        306.60      1  
7042       6844.50      0  

[7032 rows x 20 columns]

In [20]:
# One-Hot Encoding para variables con más de 2 categorías

df = pd.get_dummies(df, drop_first=True)
df

gender  SeniorCitizen  Partner  Dependents  tenure  PhoneService  \
0          1              0        1           0       1             0   
1          0              0        0           0      34             1   
2          0              0        0           0       2             1   
3          0              0        0           0      45             0   
4          1              0        0           0       2             1   
...      ...            ...      ...         ...     ...           ...   
7038       0              0        1           1      24             1   
7039       1              0        1           1      72             1   
7040       1              0        1           1      11             0   
7041       0              1        1           0       4             1   
7042       0              0        0           0      66             1   

      PaperlessBilling  MonthlyCharges  TotalCharges  Churn  ...  \
0                    1           29.85         29.85      0  ...   
1                    0           56.95       1889.50      0  ...   
2                    1           53.85        108.15      1  ...   
3                    0           42.30       1840.75      0  ...   
4                    1           70.70        151.65      1  ...   
...                ...             ...           ...    ...  ...   
7038                 1           84.80       1990.50      0  ...   
7039                 1          103.20       7362.90      0  ...   
7040                 1           29.60        346.45      0  ...   
7041                 1           74.40        306.60      1  ...   
7042                 1          105.65       6844.50      0  ...   

      TechSupport_Yes  StreamingTV_No internet service  StreamingTV_Yes  \
0               False                            False            False   
1               False                            False            False   
2               False                            False            False   
3                True                            False            False   
4               False                            False            False   
...               ...                              ...              ...   
7038             True                            False             True   
7039            False                            False             True   
7040            False                            False            False   
7041            False                            False            False   
7042             True                            False             True   

      StreamingMovies_No internet service  StreamingMovies_Yes  \
0                                   False                False   
1                                   False                False   
2                                   False                False   
3                                   False                False   
4                                   False                False   
...                                   ...                  ...   
7038                                False                 True   
7039                                False                 True   
7040                                False                False   
7041                                False                False   
7042                                False                 True   

      Contract_One year  Contract_Two year  \
0                 False              False   
1                  True              False   
2                 False              False   
3                  True              False   
4                 False              False   
...                 ...                ...   
7038               True              False   
7039               True              False   
7040              False              False   
7041              False              False   
7042              False               True   

      PaymentMethod_Credit card (automatic)  PaymentM

In [22]:
df.columns

Index(['gender', 'SeniorCitizen', 'Partner', 'Dependents', 'tenure',
       'PhoneService', 'PaperlessBilling', 'MonthlyCharges', 'TotalCharges',
       'Churn', 'MultipleLines_No phone service', 'MultipleLines_Yes',
       'InternetService_Fiber optic', 'InternetService_No',
       'OnlineSecurity_No internet service', 'OnlineSecurity_Yes',
       'OnlineBackup_No internet service', 'OnlineBackup_Yes',
       'DeviceProtection_No internet service', 'DeviceProtection_Yes',
       'TechSupport_No internet service', 'TechSupport_Yes',
       'StreamingTV_No internet service', 'StreamingTV_Yes',
       'StreamingMovies_No internet service', 'StreamingMovies_Yes',
       'Contract_One year', 'Contract_Two year',
       'PaymentMethod_Credit card (automatic)',
       'PaymentMethod_Electronic check', 'PaymentMethod_Mailed check'],
      dtype='object')

In [24]:
df.dtypes

gender                                     int64
SeniorCitizen                              int64
Partner                                    int64
Dependents                                 int64
tenure                                     int64
PhoneService                               int64
PaperlessBilling                           int64
MonthlyCharges                           float64
TotalCharges                             float64
Churn                                      int64
MultipleLines_No phone service              bool
MultipleLines_Yes                           bool
InternetService_Fiber optic                 bool
InternetService_No                          bool
OnlineSecurity_No internet service          bool
OnlineSecurity_Yes                          bool
OnlineBackup_No internet service            bool
OnlineBackup_Yes                            bool
DeviceProtection_No internet service        bool
DeviceProtection_Yes                        bool
TechSupport_No inter

In [26]:
# Las nuevas columnas estan como valores booleanos (True/False) por buenas practicas lo cambiamenos a entero (1/0)
df = df.astype(int)
df

gender  SeniorCitizen  Partner  Dependents  tenure  PhoneService  \
0          1              0        1           0       1             0   
1          0              0        0           0      34             1   
2          0              0        0           0       2             1   
3          0              0        0           0      45             0   
4          1              0        0           0       2             1   
...      ...            ...      ...         ...     ...           ...   
7038       0              0        1           1      24             1   
7039       1              0        1           1      72             1   
7040       1              0        1           1      11             0   
7041       0              1        1           0       4             1   
7042       0              0        0           0      66             1   

      PaperlessBilling  MonthlyCharges  TotalCharges  Churn  ...  \
0                    1              29            29      0  ...   
1                    0              56          1889      0  ...   
2                    1              53           108      1  ...   
3                    0              42          1840      0  ...   
4                    1              70           151      1  ...   
...                ...             ...           ...    ...  ...   
7038                 1              84          1990      0  ...   
7039                 1             103          7362      0  ...   
7040                 1              29           346      0  ...   
7041                 1              74           306      1  ...   
7042                 1             105          6844      0  ...   

      TechSupport_Yes  StreamingTV_No internet service  StreamingTV_Yes  \
0                   0                                0                0   
1                   0                                0                0   
2                   0                                0                0   
3                   1                                0                0   
4                   0                                0                0   
...               ...                              ...              ...   
7038                1                                0                1   
7039                0                                0                1   
7040                0                                0                0   
7041                0                                0                0   
7042                1                                0                1   

      StreamingMovies_No internet service  StreamingMovies_Yes  \
0                                       0                    0   
1                                       0                    0   
2                                       0                    0   
3                                       0                    0   
4                                       0                    0   
...                                   ...                  ...   
7038                                    0                    1   
7039                                    0                    1   
7040                                    0                    0   
7041                                    0                    0   
7042                                    0                    1   

      Contract_One year  Contract_Two year  \
0                     0                  0   
1                     1                  0   
2                     0                  0   
3                     1                  0   
4                     0                  0   
...                 ...                ...   
7038                  1                  0   
7039                  1                  0   
7040                  0                  0   
7041                  0                  0   
7042                  0                  1   

      PaymentMethod_Credit card (automatic)  PaymentM

In [28]:
df.dtypes

gender                                   int32
SeniorCitizen                            int32
Partner                                  int32
Dependents                               int32
tenure                                   int32
PhoneService                             int32
PaperlessBilling                         int32
MonthlyCharges                           int32
TotalCharges                             int32
Churn                                    int32
MultipleLines_No phone service           int32
MultipleLines_Yes                        int32
InternetService_Fiber optic              int32
InternetService_No                       int32
OnlineSecurity_No internet service       int32
OnlineSecurity_Yes                       int32
OnlineBackup_No internet service         int32
OnlineBackup_Yes                         int32
DeviceProtection_No internet service     int32
DeviceProtection_Yes                     int32
TechSupport_No internet service          int32
TechSupport_Y

In [30]:
# d-. Normalización (también llamada escalado Min-Max) de variables numéricas

from sklearn.preprocessing import MinMaxScaler

# Crear el escalador
scaler = MinMaxScaler()

# Definir las columnas a escalar
columnas_a_escalar = ['tenure', 'MonthlyCharges', 'TotalCharges']

# Aplicar el escalado a esas columnas
df[columnas_a_escalar] = scaler.fit_transform(df[columnas_a_escalar])

# Verificar los resultados con describe()
df[columnas_a_escalar].describe()

tenure  MonthlyCharges  TotalCharges
count  7032.000000     7032.000000   7032.000000
mean      0.442560        0.463325      0.261347
std       0.345708        0.300887      0.261570
min       0.000000        0.000000      0.000000
25%       0.112676        0.170000      0.044196
50%       0.394366        0.520000      0.159128
75%       0.760563        0.710000      0.435755
max       1.000000        1.000000      1.000000

In [31]:
# e-. Separar la variable objetivo (Churn)

# Necesitamos dividir el DataFrame en:

# X → Todas las columnas excepto la variable que queremos predecir

# y → La columna que queremos predecir: Churn (0 = no se fue, 1 = se fue)

In [32]:
# Separar variable objetivo (y) y variables predictoras (X)

X = df.drop('Churn', axis=1)
y = df['Churn']

In [36]:
# Para verificar que se la separación fue correcta

print(X.shape)
print(y.shape)
print(y.value_counts())


(7032, 30)
(7032,)
Churn
0    5163
1    1869
Name: count, dtype: int64


## 3. Dividir los datos en conjuntos de entrenamiento y prueba.

In [39]:
# Entrenamiento (train): 80%
# Prueba (test): 20%

from sklearn.model_selection import train_test_split

# División 80% entrenamiento, 20% prueba

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)

# Verifiquemos los tamaños 

print(X_train.shape)
print(X_test.shape)
print(y_train.value_counts(normalize=True))
print(y_test.value_counts(normalize=True))

(5625, 30)
(1407, 30)
Churn
0    0.734222
1    0.265778
Name: proportion, dtype: float64
Churn
0    0.734186
1    0.265814
Name: proportion, dtype: float64


## 4. Entrenar al menos 3 algoritmos y optimizar sus hiperparametros

Modelos que vamos a usar:

**1.**  Árbol de decisión (DecisionTreeClassifier)

**2.**  Regresión logística

**3.**  Naive Bayes

In [43]:
# 1. Árbol de decisión (DecisionTreeClassifier)

# Paso 1: Importar e instanciar el modelo:

from sklearn.tree import DecisionTreeClassifier

# Crear el modelo con parámetros por defecto
arbol = DecisionTreeClassifier(random_state=42)

# Entrenar el modelo con los datos de entrenamiento
arbol.fit(X_train, y_train)


DecisionTreeClassifier(random_state=42)

In [45]:
# Paso 2: Evaluar rendimiento inicial

from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

# Predecir en los datos de prueba
y_pred = arbol.predict(X_test)

# Evaluar el desempeño
print("Accuracy:", accuracy_score(y_test, y_pred))
print("\nMatriz de Confusión:\n", confusion_matrix(y_test, y_pred))
print("\nReporte de Clasificación:\n", classification_report(y_test, y_pred))


Accuracy: 0.7235252309879175

Matriz de Confusión:
 [[839 194]
 [195 179]]

Reporte de Clasificación:
               precision    recall  f1-score   support

           0       0.81      0.81      0.81      1033
           1       0.48      0.48      0.48       374

    accuracy                           0.72      1407
   macro avg       0.65      0.65      0.65      1407
weighted avg       0.72      0.72      0.72      1407



**Análisis del modelo: Árbol de Decisión**
 
Se entrenó un modelo de Árbol de Decisión utilizando los datos preprocesados para predecir la variable objetivo Churn (fuga de clientes). El modelo fue entrenado con un 80% de los datos y evaluado con el 20% restante, manteniendo el balance de clases mediante estratificación.

El modelo base, entrenado con parámetros por defecto, obtuvo un accuracy del 72.35% en el conjunto de prueba. Sin embargo, al analizar las métricas por clase, se observa un desempeño desigual:

Para la clase "0" (clientes que no se fueron), el modelo mostró un buen rendimiento con una precisión y recall del 81%.

Para la clase "1" (clientes que sí se fueron), el modelo alcanzó solo un 48% de precisión y 48% de recall, indicando dificultad para detectar correctamente los casos de fuga.

La matriz de confusión muestra una alta cantidad de falsos negativos (clientes que realmente se fueron pero el modelo predijo que no), lo cual es crítico desde el punto de vista del negocio, ya que impide tomar acciones preventivas con esos clientes.

Este comportamiento se explica, en parte, por un ligero desbalance en las clases (73% de clientes que no se fueron vs. 27% que sí). Aunque el accuracy general es aceptable, las métricas específicas para la clase minoritaria revelan la necesidad de optimizar los hiperparámetros del modelo para mejorar su capacidad de predicción sobre los casos de fuga.

In [48]:
# Vamos a Optimizar los hiperparámetros usando GridSearchCV

# Paso 1: Definir la grilla de hiperparámetros a probar

from sklearn.model_selection import GridSearchCV
from sklearn.tree import DecisionTreeClassifier

# Definir el modelo base
arbol = DecisionTreeClassifier(random_state=42)

# Definir el conjunto de hiperparámetros a probar
parametros_grid = {
    'max_depth': [3, 5, 10, None],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4],
    'criterion': ['gini', 'entropy']
}


In [50]:
# Paso 2: Configurar y ejecutar GridSearchCV

from sklearn.model_selection import GridSearchCV
from sklearn.tree import DecisionTreeClassifier

# 1. Crear el modelo base (sin entrenar todavía)
modelo_arbol = DecisionTreeClassifier(random_state=42)

# 2. Crear el objeto GridSearchCV
grid_search = GridSearchCV(
    estimator=modelo_arbol,            # el modelo que vamos a optimizar
    param_grid=parametros_grid,        # los hiperparámetros a probar
    cv=5,                              # validación cruzada con 5 particiones
    scoring='f1',                      # métrica a optimizar (se puede usar 'accuracy', 'recall', etc.)
    n_jobs=-1,                         # usa todos los núcleos disponibles (más rápido)
    verbose=1                          # muestra el progreso durante la búsqueda
)

# 3. Ejecutar la búsqueda
grid_search.fit(X_train, y_train)


Fitting 5 folds for each of 72 candidates, totalling 360 fits


GridSearchCV(cv=5, estimator=DecisionTreeClassifier(random_state=42), n_jobs=-1,
             param_grid={'criterion': ['gini', 'entropy'],
                         'max_depth': [3, 5, 10, None],
                         'min_samples_leaf': [1, 2, 4],
                         'min_samples_split': [2, 5, 10]},
             scoring='f1', verbose=1)

In [51]:
# Mostrar los mejores hiperparámetros encontrados

print("Mejores hiperparámetros:")
print(grid_search.best_params_)
print("Mejor puntuación de F1:", grid_search.best_score_)


Mejores hiperparámetros:
{'criterion': 'entropy', 'max_depth': 3, 'min_samples_leaf': 1, 'min_samples_split': 2}
Mejor puntuación de F1: 0.5852268719799805


**Análisis – Optimización del Árbol de Decisión**

Tras entrenar un modelo base de Árbol de Decisión, se procedió a optimizar sus hiperparámetros utilizando la técnica de búsqueda en grilla (GridSearchCV) con validación cruzada de 5 pliegues. Esta estrategia permitió evaluar 72 combinaciones posibles de hiperparámetros con el objetivo de maximizar el F1-score, una métrica que equilibra precisión y recall, especialmente relevante en contextos de clases desbalanceadas como el presente caso (Churn).

El modelo óptimo encontrado por GridSearchCV utiliza los siguientes parámetros:

**{'criterion': 'entropy', 'max_depth': 3, 'min_samples_leaf': 1, 'min_samples_split': 2}**

Este modelo logró un F1-score promedio de 0.585 durante la validación cruzada, lo cual representa una mejora significativa respecto al modelo base (que obtuvo un F1-score de 0.48 para la clase positiva).

La elección de una profundidad baja (3 niveles) sugiere que un árbol más simple generaliza mejor en este conjunto de datos, evitando el sobreajuste y enfocándose en patrones generales que permiten identificar con mayor efectividad a los clientes en riesgo de fuga.

In [55]:
# 2. Regresión logística

#  Paso 1: Entrenar un modelo de regresión logística básico

# Primero vamos a entrenar el modelo con los datos de entrenamiento (X_train, y_train) sin optimización, para ver cómo se comporta

from sklearn.linear_model import LogisticRegression

# Crear el modelo
modelo_logistico = LogisticRegression(random_state=42, max_iter=1000)

# Entrenar el modelo
modelo_logistico.fit(X_train, y_train)


LogisticRegression(max_iter=1000, random_state=42)

In [57]:
# Paso 2: Evaluar el rendimiento del modelo

# Calcularemos:

# Accuracy (precisión global)

# Matriz de confusión

# Reporte de clasificación (precision, recall, f1-score)

from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# Predecir en los datos de prueba
y_pred_log = modelo_logistico.predict(X_test)

# Métricas
print("Accuracy:", accuracy_score(y_test, y_pred_log))
print("\nMatriz de Confusión:\n", confusion_matrix(y_test, y_pred_log))
print("\nReporte de Clasificación:\n", classification_report(y_test, y_pred_log))


Accuracy: 0.8031272210376688

Matriz de Confusión:
 [[917 116]
 [161 213]]

Reporte de Clasificación:
               precision    recall  f1-score   support

           0       0.85      0.89      0.87      1033
           1       0.65      0.57      0.61       374

    accuracy                           0.80      1407
   macro avg       0.75      0.73      0.74      1407
weighted avg       0.80      0.80      0.80      1407



**Conclusión**

El modelo de regresión logística alcanzó una precisión del 80%, mostrando buen desempeño general. Destaca su capacidad para predecir correctamente a los clientes que no se fugan (recall 89%), aunque presenta un rendimiento más moderado en detectar los clientes en riesgo de churn (recall 57%). Esto indica que el modelo es útil, pero podría beneficiarse de una optimización adicional o técnicas para manejar el desbalance de clases.

In [60]:
# Vamos a optimización de hiperparámetros para regresión logística

# Paso 1: Definir los hiperparámetros a probar

from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import GridSearchCV

# Definir el modelo base
log_model = LogisticRegression(max_iter=1000, random_state=42)

# Definir el grid de hiperparámetros
param_grid = {
    'penalty': ['l1', 'l2', 'elasticnet', None],
    'C': [0.01, 0.1, 1, 10],  # Inverso de la regularización (entre más bajo, más regulariza)
    'solver': ['liblinear', 'saga'],  # Solvers que soportan diferentes penalties
    'l1_ratio': [None, 0.5]  # Solo aplica si penalty='elasticnet'
}


In [62]:
# Paso 2: Crear el GridSearchCV

grid_log = GridSearchCV(
    estimator=log_model,
    param_grid=param_grid,
    scoring='f1',
    cv=5,
    n_jobs=-1
)


In [66]:
# Paso 3: Ajustar el modelo

grid_log.fit(X_train, y_train)


C:\Users\Niscar\anaconda3\Lib\site-packages\sklearn\model_selection\_validation.py:540: FitFailedWarning: 
100 fits failed out of a total of 320.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
40 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\Niscar\anaconda3\Lib\site-packages\sklearn\model_selection\_validation.py", line 888, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\Niscar\anaconda3\Lib\site-packages\sklearn\base.py", line 1473, in wrapper
    return fit_method(estimator, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Niscar\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py", line 1194, in fit
    solv

GridSearchCV(cv=5, estimator=LogisticRegression(max_iter=1000, random_state=42),
             n_jobs=-1,
             param_grid={'C': [0.01, 0.1, 1, 10], 'l1_ratio': [None, 0.5],
                         'penalty': ['l1', 'l2', 'elasticnet', None],
                         'solver': ['liblinear', 'saga']},
             scoring='f1')

In [68]:
# Paso 4: Ver los mejores hiperparámetros

print("Mejores hiperparámetros:", grid_log.best_params_)
print("Mejor F1 Score en validación cruzada:", grid_log.best_score_)


Mejores hiperparámetros: {'C': 10, 'l1_ratio': 0.5, 'penalty': 'elasticnet', 'solver': 'saga'}
Mejor F1 Score en validación cruzada: 0.5956368532641968


**Conclusión**

Luego de aplicar GridSearchCV para optimizar el modelo de Regresión Logística, se determinó que los mejores hiperparámetros fueron: C=10, penalty='elasticnet', l1_ratio=0.5 y solver='saga'. Esta configuración balancea la regularización L1 y L2, favoreciendo tanto la reducción de variables irrelevantes como el control del sobreajuste. El modelo optimizado obtuvo un F1-score promedio de 0.5956, mejorando así la capacidad de detección de clientes con probabilidad de abandonar el servicio.



In [71]:
# 3. Naive Bayes

# Paso 1: Importar el modelo y entrenarlo

# Usaremos GaussianNB, que es adecuado para variables continuas

from sklearn.naive_bayes import GaussianNB

# Crear una instancia del modelo
nb_model = GaussianNB()

# Entrenar el modelo con los datos de entrenamiento
nb_model.fit(X_train, y_train)


GaussianNB()

In [73]:
# Paso 2: Evaluar el modelo en el conjunto de prueba

from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

# Predecir en los datos de prueba
y_pred_nb = nb_model.predict(X_test)

# Evaluar el modelo
print("Accuracy:", accuracy_score(y_test, y_pred_nb))
print("\nMatriz de Confusión:\n", confusion_matrix(y_test, y_pred_nb))
print("\nReporte de Clasificación:\n", classification_report(y_test, y_pred_nb))


Accuracy: 0.644633972992182

Matriz de Confusión:
 [[584 449]
 [ 51 323]]

Reporte de Clasificación:
               precision    recall  f1-score   support

           0       0.92      0.57      0.70      1033
           1       0.42      0.86      0.56       374

    accuracy                           0.64      1407
   macro avg       0.67      0.71      0.63      1407
weighted avg       0.79      0.64      0.66      1407



**Conclusión - Naive Bayes**

En el caso del modelo Naive Bayes (GaussianNB), se decidió no realizar una optimización de hiperparámetros, ya que este algoritmo cuenta con muy pocos parámetros ajustables. Su principal parámetro, var_smoothing, generalmente no produce cambios significativos en el rendimiento al ser modificado, y su impacto es menor comparado con modelos más complejos.

Además, el modelo mostró un recall elevado para la clase positiva (churn = 1), lo que indica que es útil para detectar la mayoría de los clientes que abandonan, a pesar de tener una precisión baja. Dado que el objetivo del análisis incluye identificar posibles abandonos de clientes, este comportamiento es aceptable para ciertos contextos de negocio.

Por estas razones, se considera que la optimización de hiperparámetros en este caso no aportaría mejoras relevantes y se procede a comparar su desempeño con los otros modelos evaluados.


**Comparación de Modelos**


| Modelo              | Accuracy | Precision (Clase 1) | Recall (Clase 1) | F1-Score (Clase 1) | Comentario Clave                                                                                    |
| ------------------- | -------- | ------------------- | ---------------- | ------------------ | --------------------------------------------------------------------------------------------------- |
| Árbol de Decisión   | 0.72     | 0.48                | 0.48             | 0.48               | Bajo desempeño en clase minoritaria (churn), pero balanceado en ambas clases.                       |
| Regresión Logística | 0.80     | 0.65                | 0.57             | 0.61               | Buen equilibrio entre precisión y recall. Mejor desempeño general.                                  |
| Naive Bayes (GNB)   | 0.64     | 0.42                | 0.86             | 0.56               | Excelente recall para churn, pero muy baja precisión, lo que puede generar muchos falsos positivos. |


# 5. Evaluar el rendimiento de los modelos utilizando métricas de evaluación pertinentes, como precisión, sensibilidad, valor F1, entre otras.

Evaluaremos primero **Arbol de decisión**

In [86]:
# 1. Hacer predicciones en el conjunto de prueba

mejor_modelo_arbol = grid_search.best_estimator_

y_pred_dt = mejor_modelo_arbol.predict(X_test)

# Ver métricas

from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# Evaluar el rendimiento del modelo en el conjunto de prueba
print("Accuracy:", accuracy_score(y_test, y_pred_dt))
print("\nMatriz de Confusión:\n", confusion_matrix(y_test, y_pred_dt))
print("\nReporte de Clasificación:\n", classification_report(y_test, y_pred_dt))

Accuracy: 0.751954513148543

Matriz de Confusión:
 [[829 204]
 [145 229]]

Reporte de Clasificación:
               precision    recall  f1-score   support

           0       0.85      0.80      0.83      1033
           1       0.53      0.61      0.57       374

    accuracy                           0.75      1407
   macro avg       0.69      0.71      0.70      1407
weighted avg       0.77      0.75      0.76      1407



**Conclusión - Arbol de decisión**

El modelo mejoró su rendimiento respecto al árbol base, especialmente en la clase 1 (clientes que se fueron).

Tiene una accuracy decente (75%), y un F1-score razonable en la clase minoritaria.

Aún comete errores, especialmente al confundir clientes que no se irán (204 Falsos Positivos), pero es un comportamiento típico cuando las clases están desbalanceadas.

Evaluaremos  segundo **Regresión Logística**

In [90]:
# Paso 1:  Mejor modelo optimizado

mejor_modelo_log = grid_search.best_estimator_

# Paso 2: Realizar predicciones en el conjunto de prueba

y_pred_log = mejor_modelo_log.predict(X_test)

# Paso 3: Evaluar el modelo con métricas

from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Accuracy
accuracy_log = accuracy_score(y_test, y_pred_log)
print("Accuracy:", accuracy_log)

# Matriz de confusión
print("Matriz de Confusión:\n", confusion_matrix(y_test, y_pred_log))

# Reporte de clasificación
print("Reporte de Clasificación:\n", classification_report(y_test, y_pred_log))


Accuracy: 0.751954513148543
Matriz de Confusión:
 [[829 204]
 [145 229]]
Reporte de Clasificación:
               precision    recall  f1-score   support

           0       0.85      0.80      0.83      1033
           1       0.53      0.61      0.57       374

    accuracy                           0.75      1407
   macro avg       0.69      0.71      0.70      1407
weighted avg       0.77      0.75      0.76      1407



**Conclusión**

El modelo identifica mejor que el árbol de decisión a los clientes en riesgo de churn, con mejor F1 para clase 1.

Aunque aún tiene margen de mejora, es más equilibrado en su rendimiento.

El recall para churn (0.61) es importante en negocios, ya que prefieres detectar más posibles fugas, aunque algunos falsos positivos se filtren.



Evaluaremos  tercero **Naive Bayes**

In [102]:

from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# 1. Entrenar el modelo 

modelo_nb = GaussianNB()
modelo_nb.fit(X_train, y_train)

# 2. Predecir con los datos de prueba

y_pred_nb = modelo_nb.predict(X_test)

# 3. Evaluar el modelo

print("Accuracy:", accuracy_score(y_test, y_pred_nb))
print("\nMatriz de Confusión:\n", confusion_matrix(y_test, y_pred_nb))
print("\nReporte de Clasificación:\n", classification_report(y_test, y_pred_nb))


Accuracy: 0.644633972992182

Matriz de Confusión:
 [[584 449]
 [ 51 323]]

Reporte de Clasificación:
               precision    recall  f1-score   support

           0       0.92      0.57      0.70      1033
           1       0.42      0.86      0.56       374

    accuracy                           0.64      1407
   macro avg       0.67      0.71      0.63      1407
weighted avg       0.79      0.64      0.66      1407



 **Conclusión: ¿Cuál es el mejor modelo?**

| Modelo              | Accuracy | F1-score clase 1 | Observación clave                  |
| ------------------- | -------- | ---------------- | ---------------------------------- |
| Árbol de decisión   | 0.75     | 0.57             | Buen equilibrio y F1-score decente |
| Regresión logística | 0.80     | 0.61             | El **mejor desempeño general**     |
| Naive Bayes         | 0.64     | 0.56             | Alto recall, pero baja precisión   |


**Conclusión final del análisis de modelos**

Luego de entrenar, optimizar y evaluar tres modelos de clasificación (Árbol de Decisión, Regresión Logística y Naive Bayes), se concluye que la Regresión Logística fue el modelo con mejor desempeño general. Este modelo alcanzó una precisión del 80% y un F1-score de 0.61 para la clase positiva (Churn), lo que indica un buen equilibrio entre precisión y recall, especialmente relevante en contextos con clases desbalanceadas.

El modelo de Árbol de Decisión, aunque ligeramente inferior en desempeño (F1-score = 0.57 para la clase 1), también mostró buenos resultados y es más interpretable, lo cual podría ser útil en escenarios donde la explicación del modelo es importante.

Por su parte, el modelo de Naive Bayes se destacó por su alto recall (0.86) en la clase de clientes que abandonan, lo cual puede ser ventajoso si se prioriza la detección de posibles bajas. Sin embargo, su baja precisión (0.42) y su accuracy general de 64% lo hacen menos confiable para una predicción equilibrada.

En conclusión, la Regresión Logística se presenta como la mejor opción para este problema de clasificación de Churn, al combinar una alta precisión global con una buena capacidad de detección de clientes en riesgo.